### PR TITLE
feat: add onBeforeMoveEnd

### PR DESCRIPTION
**Pull request by linxianxi**
issue: #318

### CODE REVIEW

File: packages/core/src/interfaces/editor.ts [hunk 1]
```diff
@@ -10,6 +10,7 @@ import { CoreEventHandlers } from '../events';
 
 export type Options = {
   onRender: React.ComponentType<{ render: React.ReactElement }>;
+  onBeforeMoveEnd: (targetNode: Node, currentNode: Node) => void;
```
Comment (prevwong) [inline packages/core/src/interfaces/editor.ts:13]: Based on the function call in the `actions` file above, this should be `onBeforeMoveEnd: (newParentNode: Node, targetNode: Node)`, right? 

Perhaps we should also include the existingParentNode as well? So something like: `onBeforeMoveEnd: (targetNode: Node, newParentNode: Node, existingParentNode: Node)`. This way related to your issue #318, when a user moves a Node, you can update the width of the Node that was moved, and you can also update the widths of the children of the existing and new parent node.

**Review by prevwong — COMMENTED** [review]
Thanks for this PR! Think this option could be quite handy. Could you also add some tests here: https://github.com/prevwong/craft.js/blob/next/packages/core/src/editor/tests/actions.test.ts#L447

File: packages/core/src/editor/Editor.tsx [hunk 2]
```diff
@@ -30,7 +31,14 @@ export const Editor: React.FC<Partial<Options>> = ({
...
+        onRender,
+        onNodesChange,
+        onBeforeMoveEnd,
+        resolver,
+        enabled,
+        indicator,
+      },
       (value) => value !== undefined
     );
   }, [enabled, indicator, onNodesChange, onRender, resolver]);
```
Comment (ankri) [inline packages/core/src/editor/Editor.tsx:44]: Does `onBeforeMoveEnd` need to be a part of the `useMemo` dependency array?

**Review by ankri — COMMENTED** [review]
I like @prevwong's suggestion. Also please check if `onBeforeMoveEnd` needs to be part of the `useMemo` dependency array.

**Comment by linxianxi** [comment]
@prevwong  Please review again.

**Review by ankri — APPROVED** [review]

**Review by prevwong — APPROVED** [review]
LGTM 💯  Thanks!

**Pull request merged**